cabal-install/Distribution/Client/Dependency/Modular/Cycles.hs

```diff
@@ -7,33 +7,32 @@ import Prelude hiding (cycle)
 import Control.Monad
 import Control.Monad.Reader
 import Data.Graph (SCC)
-import Data.Set   (Set)
+import Data.Map   (Map)
 import qualified Data.Graph       as Gr
 import qualified Data.Map         as Map
-import qualified Data.Set         as Set
 import qualified Data.Traversable as T
 
 #if !MIN_VERSION_base(4,8,0)
 import Control.Applicative ((<$>))
 #endif
 
 import Distribution.Client.Dependency.Modular.Dependency
-import Distribution.Client.Dependency.Modular.Flag
 import Distribution.Client.Dependency.Modular.Package
 import Distribution.Client.Dependency.Modular.Tree
+import qualified Distribution.Client.Dependency.Modular.ConflictSet as CS
 
-type DetectCycles = Reader (ConflictSet QPN)
+type DetectCycles = Reader (Map QPN QGoalReason)
```

Comment:
Ok, I know we talked about this already and I said to do it later, but looking at it now, I agree with you. Collecting all these `QGoalReason`s now for no purpose is not worth it. We really don't need any `Reader` here now, right? Just the `RevDepMap` in the end?